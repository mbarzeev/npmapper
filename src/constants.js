const cwd = process.cwd();

export const HTML_REPORT_TEMPLATE_PATH = `${__dirname}/reports/html/report-template.html`;
export const REPORTS_OUTPUT_PATH = `${cwd}/npmapper`;
export const REPORT_FILE_NAME = 'npmapper-report';
export const TYPE_SCRIPT = 'script';
export const TYPE_COMMAND = 'command';
export const NOT_AVAILABLE = 'N/A';
export const SERIAL_COMMAND_DELIMITER = '&&';
export const PRE_SCRIPT_PREFIX = 'pre';
export const POST_SCRIPT_PREFIX = 'post';
export const NESTED_SCRIPT_REGEX = /^npm run\s/;
export const PREFIX_REGEX = /--prefix /;
export const PREFIX_VALUE_REGEX = /--prefix\s+("[^"]*"|\S*)/;
export const CONFIG_PROP_REGEX = /\$npm_package_config_/;
export const COMMAND_NAME_REGEX = /.*?(?=\s-)/;
export const PARAMS_REGEX = new RegExp(`(?<=\\s)-(?:(?!(\\s-|${SERIAL_COMMAND_DELIMITER})).)*`, 'gm');
export const EQUAL_SIGN_REGEX = /\=/;
export const NPM_PARAMS = /(--silent|--quiet)/;

export const SCRIPT_ACTION_METADATA = {
    title: 'NPM Script',
    iconClass: 'icofont-gear',
};

export const COMMAND_ACTION_METADATA = {
    title: 'Command',
    iconClass: 'icofont-automation',
};

export const ACTION_METADATA_DICT = {
    [TYPE_SCRIPT]: SCRIPT_ACTION_METADATA,
    [TYPE_COMMAND]: COMMAND_ACTION_METADATA,
};

export const HTML_REPORT_TYPE = 'html';
export const JSON_REPORT_TYPE = 'json';
