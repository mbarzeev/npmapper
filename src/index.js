import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import reduce from 'awaity/reduce';
import open from 'open';
import {generateReport as generateHtmlReport} from './reports/html/html-reporter';
import {
    REPORTS_OUTPUT_PATH,
    REPORT_FILE_NAME,
    TYPE_SCRIPT,
    TYPE_COMMAND,
    HTML_REPORT_TYPE,
    JSON_REPORT_TYPE,
    NOT_AVAILABLE,
    SERIAL_COMMAND_DELIMITER,
    PRE_SCRIPT_PREFIX,
    POST_SCRIPT_PREFIX,
    NESTED_SCRIPT_REGEX,
    PREFIX_REGEX,
    PREFIX_VALUE_REGEX,
    CONFIG_PROP_REGEX,
    COMMAND_NAME_REGEX,
    PARAMS_REGEX,
    EQUAL_SIGN_REGEX,
    NPM_PARAMS,
} from './constants';

let originJsonFilePath;

/**
 * The tool's entry point. Extracts the CLI params, calls to map the scripts and
 * creates a report
 * @param {Object} options - An object containing the params passed on the CLI
 */
export async function mapNpmScripts(options) {
    const {targetDir, scriptName, reportType} = options;
    const jsonFilePath = path.resolve(targetDir, 'package.json');
    try {
        originJsonFilePath = jsonFilePath;
        const result = await parseNpmScripts(jsonFilePath, scriptName);
        createReport(result, targetDir, reportType);
    } catch (error) {
        console.error(`${chalk.red.bold(error)}`);
        throw error;
    }
}

/**
 * Reads the package.json file content and calls to retrieve the mapping result
 * from it
 * @param {string} jsonFilePath - The path to the package.json file
 * @param {string} desiredScript - A specific script we wish to map
 * @returns {Array} The mapping result Array
 */
export async function parseNpmScripts(jsonFilePath, desiredScript) {
    // Check if the file is accessible
    await getIsJsonFileAccessible(jsonFilePath);
    // Read the file
    const jsonFileContent = await getJsonFileContent(jsonFilePath);
    // Map the npm scripts
    const scripts = desiredScript ? [desiredScript] : Object.keys(jsonFileContent.scripts);
    const result = await getScriptsArray(jsonFileContent, scripts);
    // If we are "jumping" to another package.json file we indicate it by
    // setting the location of the first script on that location
    if (jsonFilePath !== originJsonFilePath) {
        result[0].location = jsonFilePath;
    }
    return result;
}

/**
 * Returns an Array representing the scripts tree according to the scripts provided
 * @param {Object} jsonFileContent - An object representing the package.json content
 * @param {Array} scripts - An array of scripts to map
 * @returns {Array} An Array representing the scripts
 */
async function getScriptsArray(jsonFileContent, scripts) {
    return await reduce(
        scripts,
        async (accumulate, scriptName) => {
            const singleScriptObject = await getSingleScriptObject(scriptName, jsonFileContent);
            accumulate.push(singleScriptObject);
            return accumulate;
        },
        []
    );
}

/**
 * Returns a single script object by the given script name. This method also
 * supports recursion if there are nested scripts detected.
 * If needed, it can also "jump" to another package.json if a nested script with
 * prefix param was detected
 * @param {string} scriptName - The script name we wish to map
 * @param {Object} jsonFileContent - An object representing the package.json content
 * @returns {Object} A single script object with all the data gathered for it
 * @throws Will throw an error if the script cannot be found
 */
async function getSingleScriptObject(scriptName, jsonFileContent) {
    const scriptsConfig = jsonFileContent.config;

    // If the scriptName is actually an action which has a prefix parameter to it...
    if (PREFIX_REGEX.test(scriptName)) {
        const prefixValue = PREFIX_VALUE_REGEX.exec(scriptName)[1];
        let prefixPath;

        // If the config value is resolved from an npm config extract the value
        // from it otherwise take it as is
        if (CONFIG_PROP_REGEX.test(prefixValue)) {
            const configProp = prefixValue.replace(CONFIG_PROP_REGEX, '');
            prefixPath = scriptsConfig[configProp];
        } else {
            prefixPath = prefixValue;
        }

        // Remove all the command params and spaces from the nested script
        const trimmedScriptName = scriptName.replace(PARAMS_REGEX, '').trim();
        const nestedScriptTree = await parseNpmScripts(path.resolve(prefixPath, 'package.json'), trimmedScriptName);
        return nestedScriptTree[0];
    } else {
        // Does the script have params to it?
        const hasParams = PARAMS_REGEX.test(scriptName);
        let params = [];

        if (hasParams) {
            params = extractParamsFromActionStr(scriptName, scriptsConfig);
            scriptName = COMMAND_NAME_REGEX.exec(scriptName)[0];
        }

        const trimmedScriptName = scriptName && scriptName.trim();
        const scriptsObj = jsonFileContent.scripts;

        if (trimmedScriptName && scriptsObj[trimmedScriptName]) {
            const rawScriptObj = scriptsObj[trimmedScriptName];
            const actions = rawScriptObj.split(SERIAL_COMMAND_DELIMITER).map(step => step.trim());
            const actionsArray = await getActionsArray(actions, jsonFileContent);
            const resultScriptObj = {
                type: TYPE_SCRIPT,
                name: scriptName,
                pre: NOT_AVAILABLE,
                post: NOT_AVAILABLE,
                params,
                actions: [],
            };

            // If the script has a pre script
            if (scriptsObj[`${PRE_SCRIPT_PREFIX}${trimmedScriptName}`]) {
                resultScriptObj.pre = await getSingleScriptObject(
                    `${PRE_SCRIPT_PREFIX}${trimmedScriptName}`,
                    jsonFileContent
                );
            }

            // The script itself
            resultScriptObj.actions = actionsArray;

            // If the script has a post script
            if (scriptsObj[`${POST_SCRIPT_PREFIX}${trimmedScriptName}`]) {
                resultScriptObj.post = await getSingleScriptObject(
                    `${POST_SCRIPT_PREFIX}${trimmedScriptName}`,
                    jsonFileContent
                );
            }

            return resultScriptObj;
        } else {
            throw new Error(`No script by the name "${trimmedScriptName}" was found`);
        }
    }
}

/**
 * Given an action (script or command) string, it will attempt to extract the
 * params from it and return as a structured object
 * @param {String} actionStr - The full action string with params
 * @param {Object} scriptsConfig - An object representing the npm configuration
 * @returns {Array} The processed params array
 */
function extractParamsFromActionStr(actionStr, scriptsConfig = {}) {
    const totalFullParams = actionStr.match(PARAMS_REGEX);
    const params = totalFullParams.map(singleFullParam => {
        const isSeparateWithEqualSign = EQUAL_SIGN_REGEX.test(singleFullParam);
        const separator = isSeparateWithEqualSign ? '=' : ' ';
        let [name, value] = singleFullParam.split(separator);
        if (CONFIG_PROP_REGEX.test(value)) {
            const configProp = value.replace(CONFIG_PROP_REGEX, '');
            value = `${scriptsConfig[configProp]} (from npm configuration named "${configProp}")`;
        }
        return {
            name,
            value,
        };
    });
    return params;
}

/**
 * Goes over the raw actions array given and process it
 * @param {Array} rawActions
 * @param {Object} jsonFileContent - An object representing the package.json content
 * @returns {Array}
 */
async function getActionsArray(rawActions, jsonFileContent) {
    const scriptsConfig = jsonFileContent.config;
    return await reduce(
        rawActions,
        async (acc, curr) => {
            // If the action is a nested NPM script we wish to dig inside it
            if (NESTED_SCRIPT_REGEX.test(curr)) {
                const nestedScriptName = curr.replace(NESTED_SCRIPT_REGEX, '');
                // Trim and clear any NPM's params (in the future it will be nice
                // to support them as well)
                const trimmedNestedScriptName = nestedScriptName.replace(NPM_PARAMS, '').trim();
                const singleScriptObject = await getSingleScriptObject(trimmedNestedScriptName, jsonFileContent);
                acc.push(singleScriptObject);
            } else {
                // If the action is not a nested script we are dealing with a command
                acc.push(getSingleCommandObject(curr, scriptsConfig));
            }
            return acc;
        },
        []
    );
}

/**
 * Returns an Object which represents a single command with its params attached
 * @param {string} rawCommand
 * @param {Object} scriptsConfig - An object representing the npm configuration
 * @returns {Object} An Object representing a single command
 */
export function getSingleCommandObject(rawCommand, scriptsConfig) {
    let params = [];
    const hasParams = PARAMS_REGEX.test(rawCommand);

    if (hasParams) {
        params = extractParamsFromActionStr(rawCommand, scriptsConfig);
        rawCommand = COMMAND_NAME_REGEX.exec(rawCommand)[0];
    }
    return {type: TYPE_COMMAND, name: rawCommand, params};
}

/**
 * Returns the json file content found on the path given
 * @param {String} jsonFilePath
 */
async function getJsonFileContent(jsonFilePath) {
    try {
        const jsonFileContentData = await fs.readFile(jsonFilePath, 'utf-8');
        return JSON.parse(jsonFileContentData);
    } catch (error) {
        console.error('%s Could not read package.json file', chalk.red.bold('ERROR'));
        process.exit(1);
    }
}

/**
 * Attempts to access the json file given and exist if fails.
 * @param {String} jsonFilePath
 * @throws Will throw an error if a package.json file cannot be found in the
 * given path
 */
async function getIsJsonFileAccessible(jsonFilePath) {
    try {
        await fs.access(jsonFilePath, fs.constants.R_OK);
    } catch (error) {
        throw new Error(`No package.json file can be found on ${jsonFilePath}`);
    }
}

/**
 * Calls to generate a report according to the type requested
 * @param {Object} scriptsMappingResult - The mapping result as an Object
 * @param {string} packageJsonFilePath
 * @param {string} reportType - 'html' or 'json', default is html
 */
async function createReport(scriptsMappingResult, packageJsonFilePath, reportType) {
    switch (reportType) {
        case JSON_REPORT_TYPE:
            await createJsonReport(scriptsMappingResult);
            break;
        case HTML_REPORT_TYPE:
        default:
            await createHtmlReport(scriptsMappingResult, packageJsonFilePath);
            break;
    }
}

/**
 * Takes the scripts mapping results, converts it to a string and call to write
 * it to a file
 * @param {Object} scriptsMappingResult - The mapping
 */
async function createJsonReport(scriptsMappingResult) {
    const fileName = `${REPORT_FILE_NAME}.json`;
    writeReportFile(fileName, JSON.stringify(scriptsMappingResult));
}

/**
 * Calls to generate an HTML report from the mapping result, then copies assets
 * to support the report and opens the report in a browser
 * @param {Object} scriptsMappingResult - The mapping result as an Object
 * @param {string} packageJsonFilePath
 */
async function createHtmlReport(scriptsMappingResult, packageJsonFilePath) {
    const htmlReport = await generateHtmlReport(scriptsMappingResult, packageJsonFilePath);
    const fileName = `${REPORT_FILE_NAME}.html`;
    await writeReportFile(fileName, htmlReport);
    // Copy assets files to the target directory
    fs.copyFileSync(`${__dirname}/reports/html/report.css`, `${REPORTS_OUTPUT_PATH}/report.css`);
    fs.copyFileSync(`${__dirname}/reports/html/icofont.min.css`, `${REPORTS_OUTPUT_PATH}/icofont.min.css`);
    fs.copySync(`${__dirname}/reports/html/fonts`, `${REPORTS_OUTPUT_PATH}/fonts`);
    // Open the report in the default browser
    open(`${REPORTS_OUTPUT_PATH}/${REPORT_FILE_NAME}.html`);
}

/**
 * Writes the report to the file system using the given file name and the content
 * @param {string} fileName - The file name to be used for the report
 * @param {string} content - The content of the report
 */
async function writeReportFile(fileName, content) {
    if (!fs.existsSync(REPORTS_OUTPUT_PATH)) {
        await fs.mkdirSync(REPORTS_OUTPUT_PATH);
    }
    await fs.writeFile(`${REPORTS_OUTPUT_PATH}/${fileName}`, content);
}
