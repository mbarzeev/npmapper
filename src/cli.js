import {mapNpmScripts} from './index';
import {HTML_REPORT_TYPE} from './constants';
import program from 'commander';

program
    .option('-d, --dir <string>', 'The directory npmapper should work on (where the package.json is at)', process.cwd())
    .option('-s, --script-name <string>', 'A specific npm script you want to map')
    .option('-t, --report-type <type>', 'The desired type of the report. Default is HTML. use json to generate a json report', HTML_REPORT_TYPE);

export function cli(args) {
    program.parse(args);
    const targetDir = program.dir;
    const scriptName = program.scriptName;
    const reportType = program.reportType;
    mapNpmScripts({targetDir, scriptName, reportType});
}
