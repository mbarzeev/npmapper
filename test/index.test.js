import fs from 'fs-extra';
import path from 'path';
jest.mock('path');
jest.mock('fs');

jest.mock('../src/reports/html/html-reporter.js', () => ({
    generateReport: jest.fn().mockReturnValue('mockHtmlReportContent'),
}));

import {mapNpmScripts} from '../src/index';
import * as constants from '../src/constants';

describe('NPMapper tool - ', () => {
    const MOCK_PATH = 'mockPath';
    const jsonFileContent = `{"scripts":{"build":"mock build action", "test":"mock test action"}}`;

    beforeEach(() => {
        fs.access = jest.fn();
        fs.readFile = jest.fn();
        fs.readFile.mockReturnValue(jsonFileContent);
        console.error = jest.fn();
        path.resolve.mockImplementation(givenPath => givenPath);
    });

    describe('File availability', () => {
        it('should check if the package.json file is accessible', async () => {
            // await parseNpmScripts(MOCK_PATH);
            await mapNpmScripts({targetDir: MOCK_PATH});
            expect(fs.access).toHaveBeenCalledWith(MOCK_PATH, fs.constants.F_OK);
        });

        it('should fail is the file is not accessible', async () => {
            const ERROR_MSG = 'No package.json file can be found on mockPath';
            fs.access.mockImplementation(() => {
                throw new Error();
            });
            const throwableWrapperFunc = async () => {
                await mapNpmScripts({targetDir: MOCK_PATH});
            };
            await expect(throwableWrapperFunc()).rejects.toThrow(ERROR_MSG);
        });

        it('should attempt to read the package.json file', async () => {
            await mapNpmScripts({targetDir: MOCK_PATH});
            expect(fs.readFile).toHaveBeenCalledWith(MOCK_PATH, 'utf-8');
        });

        it('should fail if was unable to read the package.json file', async () => {
            const ERROR_MSG = 'Could not read the package.json file under mockPath';
            fs.readFile.mockImplementation(() => {
                throw new Error();
            });
            const throwableWrapperFunc = async () => {
                await mapNpmScripts({targetDir: MOCK_PATH});
            };
            await expect(throwableWrapperFunc()).rejects.toThrow(ERROR_MSG);
        });
    });

    describe('Parsing scripts', () => {
        it('should get the scripts array for a specific one', async () => {
            const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
            expect(result).toMatchSnapshot();
        });

        it('should fail if the script cannot be found on the package.json file', async () => {
            const NON_EXISTING_SCRIPT_NAME = 'nonExistingScriptName';
            const ERROR_MSG = `No script by the name "${NON_EXISTING_SCRIPT_NAME}" was found`;
            const throwableWrapperFunc = async () => {
                await mapNpmScripts({targetDir: MOCK_PATH, scriptName: NON_EXISTING_SCRIPT_NAME});
            };
            await expect(throwableWrapperFunc()).rejects.toThrow(ERROR_MSG);
        });

        it('should auto map pre scripts', async () => {
            const jsonFileContent = `
                    {
                       "scripts": {
                            "prebuild": "mock pre build script",
                            "build": "mock build script"
                        }
                    }`;
            fs.readFile.mockReturnValue(jsonFileContent);
            const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
            expect(result).toMatchSnapshot();
        });

        it('should auto map post scripts', async () => {
            const jsonFileContent = `
                    {
                       "scripts": {
                            "postbuild": "mock post build script",
                            "build": "mock build script"
                        }
                    }`;
            fs.readFile.mockReturnValue(jsonFileContent);
            const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
            expect(result).toMatchSnapshot();
        });

        it('should parse a nested npm script', async () => {
            const jsonFileContent = `
                {
                    "scripts": {
                        "build": "npm run other",
                        "other": "mock other action"
                    }
                }`;
            fs.readFile.mockReturnValue(jsonFileContent);
            const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
            expect(result[0].actions[0].name).toEqual('other');
            expect(result[0].actions[0].type).toEqual('script');
        });

        it('should parse a nested yarn script without using "run" to execute it', async () => {
            const jsonFileContent = `
                {
                    "scripts": {
                        "build": "yarn other",
                        "other": "mock other action"
                    }
                }`;
            fs.readFile.mockReturnValue(jsonFileContent);
            const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
            expect(result[0].actions[0].name).toEqual('other');
            expect(result[0].actions[0].type).toEqual('script');
        });

        it('should parse a nested yarn script with using "run" to execute it', async () => {
            const jsonFileContent = `
                {
                    "scripts": {
                        "build": "yarn run other",
                        "other": "mock other action"
                    }
                }`;
            fs.readFile.mockReturnValue(jsonFileContent);
            const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
            expect(result[0].actions[0].name).toEqual('other');
            expect(result[0].actions[0].type).toEqual('script');
        });

        describe('script parameters parsing', () => {
            it('should extract and append parameters if exist', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "npm run other --param1 value1",
                        "other": "mock other action"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should handle params which are defined with an "=" sign', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "mock script --param1=value1"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should resolve values which are defined in npm config', async () => {
                const jsonFileContent = `
                {
                    "config": {
                        "mockParamValueConfig": "mockParamValue"
                    },
                    "scripts": {
                        "build": "mock script --param1 $npm_package_config_mockParamValueConfig"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should return action object for a singe dashed parameter with value', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "mock script -param1 param1Value"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should return action object for a single dashed parameter without value', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "mock script -param1"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should return action object for a multi dashed parameter with values', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "mock script --firstParam firstParamValue --secondParam secondParamValue --thirdParam thirdParamValue"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should return action object for a multi dashed parameter some with values and some without', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "mock script --firstParam firstParamValue --secondParam --thirdParam"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            // TODO: This test does not pass well - the first param is not registered
            it('should return action object for a multi dashed parameter with values with random spaces', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "mock script --firstParam firstParamValue    --secondParam secondParamValue    --thirdParam thirdParamValue"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should return action object for a single dashed parameter with name and value separate with "=" sign', async () => {
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "mock script --firstParam=firstParamValue"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });
        });

        describe('handling "--prefix" parameter', () => {
            it('should hop to other package.json file when --prefix param is set to a value', async () => {
                // Note - for mock purposes, the same JSON file is read for both paths
                // therefore the scripts hold the "other" script as well
                const jsonFileContent = `
                {
                    "scripts": {
                        "build": "npm run other --prefix otherMockPath", 
                        "other": "mock other action"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });

            it('should hop to other package.json file when --prefix param is set to an npm configuration', async () => {
                // Note - for mock purposes, the same JSON file is read for both paths
                // therefore the scripts hold the "other" script as well
                const jsonFileContent = `
                {
                    "config": {
                        "mockPathConfig": "otherMockPath"
                    },
                    "scripts": {
                        "build": "npm run other --prefix $npm_package_config_mockPathConfig", 
                        "other": "mock other action"
                    }
                }`;
                fs.readFile.mockReturnValue(jsonFileContent);
                const result = await mapNpmScripts({targetDir: MOCK_PATH, scriptName: 'build'});
                expect(result).toMatchSnapshot();
            });
        });
    });

    describe('Reporting', () => {
        const MOCK_REPORTS_OUTPUT_PATH = 'mockReportsOutputPath';

        beforeAll(() => {
            fs.writeFile = jest.fn();
            fs.copySync = jest.fn();
            fs.existsSync = jest.fn();
            fs.mkdirSync = jest.fn();

            constants.REPORTS_OUTPUT_PATH = MOCK_REPORTS_OUTPUT_PATH;
        });

        it('should create an JSON report if requested', async () => {
            const expectedFileName = `${MOCK_REPORTS_OUTPUT_PATH}/${constants.REPORT_FILE_NAME}.json`;
            const expectedFileContent =
                '[{"type":"script","name":"build","pre":"N/A","post":"N/A","params":[],"actions":[{"type":"command","name":"mock build action","params":[]}]}]';

            await mapNpmScripts({
                targetDir: MOCK_PATH,
                scriptName: 'build',
                reportType: constants.JSON_REPORT_TYPE,
            });

            expect(fs.writeFile).toHaveBeenCalledWith(expectedFileName, expectedFileContent);
        });

        it('should create an HTML report if requested', async () => {
            const expectedFileName = `${MOCK_REPORTS_OUTPUT_PATH}/${constants.REPORT_FILE_NAME}.html`;

            await mapNpmScripts({
                targetDir: MOCK_PATH,
                scriptName: 'build',
                reportType: constants.HTML_REPORT_TYPE,
            });

            expect(fs.writeFile).toHaveBeenCalledWith(expectedFileName, 'mockHtmlReportContent');
        });

        it('should create the reports directory if it does not exist', async () => {
            fs.existsSync.mockReturnValue(false);

            await mapNpmScripts({
                targetDir: MOCK_PATH,
                scriptName: 'build',
                reportType: constants.HTML_REPORT_TYPE,
            });

            expect(fs.mkdirSync).toHaveBeenCalledWith(MOCK_REPORTS_OUTPUT_PATH);
        });

        it('should not create the reports directory if it exists', async () => {
            fs.existsSync.mockReturnValue(true);

            await mapNpmScripts({
                targetDir: MOCK_PATH,
                scriptName: 'build',
                reportType: constants.HTML_REPORT_TYPE,
            });

            expect(fs.mkdirSync).not.toHaveBeenCalled();
        });
    });
});
