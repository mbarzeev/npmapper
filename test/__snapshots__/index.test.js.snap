// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NPMapper tool -  Parsing scripts handling "--prefix" parameter should hop to other package.json file when --prefix param is set to a value 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "actions": Array [
          Object {
            "name": "mock other action",
            "params": Array [],
            "type": "command",
          },
        ],
        "location": "otherMockPath",
        "name": "other",
        "params": Array [],
        "post": "N/A",
        "pre": "N/A",
        "type": "script",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts handling "--prefix" parameter should hop to other package.json file when --prefix param is set to an npm configuration 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "actions": Array [
          Object {
            "name": "mock other action",
            "params": Array [],
            "type": "command",
          },
        ],
        "location": "otherMockPath",
        "name": "other",
        "params": Array [],
        "post": "N/A",
        "pre": "N/A",
        "type": "script",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should extract and append parameters if exist 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "actions": Array [
          Object {
            "name": "mock other action",
            "params": Array [],
            "type": "command",
          },
        ],
        "name": "other",
        "params": Array [
          Object {
            "name": "--param1",
            "value": "value1",
          },
        ],
        "post": "N/A",
        "pre": "N/A",
        "type": "script",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should handle params which are defined with an "=" sign 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "--param1",
            "value": "value1",
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should resolve values which are defined in npm config 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "--param1",
            "value": "mockParamValue (from npm configuration named \\"mockParamValueConfig\\")",
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should return action object for a multi dashed parameter some with values and some without 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "--firstParam",
            "value": "firstParamValue",
          },
          Object {
            "name": "--secondParam",
            "value": undefined,
          },
          Object {
            "name": "--thirdParam",
            "value": undefined,
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should return action object for a multi dashed parameter with values 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "--firstParam",
            "value": "firstParamValue",
          },
          Object {
            "name": "--secondParam",
            "value": "secondParamValue",
          },
          Object {
            "name": "--thirdParam",
            "value": "thirdParamValue",
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should return action object for a multi dashed parameter with values with random spaces 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "--firstParam",
            "value": "firstParamValue",
          },
          Object {
            "name": "--secondParam",
            "value": "secondParamValue",
          },
          Object {
            "name": "--thirdParam",
            "value": "thirdParamValue",
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should return action object for a singe dashed parameter with value 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "-param1",
            "value": "param1Value",
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should return action object for a single dashed parameter with name and value separate with "=" sign 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "--firstParam",
            "value": "firstParamValue",
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts script parameters parsing should return action object for a single dashed parameter without value 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock script",
        "params": Array [
          Object {
            "name": "-param1",
            "value": undefined,
          },
        ],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts should auto map post scripts 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock build script",
        "params": Array [],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": Object {
      "actions": Array [
        Object {
          "name": "mock post build script",
          "params": Array [],
          "type": "command",
        },
      ],
      "name": "postbuild",
      "params": Array [],
      "post": "N/A",
      "pre": "N/A",
      "type": "script",
    },
    "pre": "N/A",
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts should auto map pre scripts 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock build script",
        "params": Array [],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": Object {
      "actions": Array [
        Object {
          "name": "mock pre build script",
          "params": Array [],
          "type": "command",
        },
      ],
      "name": "prebuild",
      "params": Array [],
      "post": "N/A",
      "pre": "N/A",
      "type": "script",
    },
    "type": "script",
  },
]
`;

exports[`NPMapper tool -  Parsing scripts should get the scripts array for a specific one 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "name": "mock build action",
        "params": Array [],
        "type": "command",
      },
    ],
    "name": "build",
    "params": Array [],
    "post": "N/A",
    "pre": "N/A",
    "type": "script",
  },
]
`;
